File: src/main/java/org/jcodec/containers/mp4/muxer/TimecodeMP4MuxerTrack.java
package org.jcodec.containers.mp4.muxer;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

import org.jcodec.common.Codec;
import org.jcodec.common.io.SeekableByteChannel;
import org.jcodec.common.model.Packet;
import org.jcodec.common.model.Rational;
import org.jcodec.common.model.TapeTimecode;
import org.jcodec.containers.mp4.MP4Packet;
import org.jcodec.containers.mp4.MP4TrackType;
import org.jcodec.containers.mp4.boxes.Box;
import org.jcodec.containers.mp4.boxes.Edit;
import org.jcodec.containers.mp4.boxes.MovieHeaderBox;
import org.jcodec.containers.mp4.boxes.TimecodeSampleEntry;
import org.jcodec.movtool.Util;

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 * 
 * Timecode MP4 muxer track
 * 
 * @author The JCodec project
 * 
 */
public class TimecodeMP4MuxerTrack extends FramesMP4MuxerTrack {

    private TapeTimecode prevTimecode;
    private TapeTimecode firstTimecode;

    private int fpsEstimate;
    private long sampleDuration;
    private long samplePts;
    private int tcFrames;
    private List<Edit> lower;
    private List<Packet> gop;

    public TimecodeMP4MuxerTrack(SeekableByteChannel out, int trackId) {
        super(out, trackId, MP4TrackType.TIMECODE, Codec.TIMECODE);
        this.lower = new ArrayList<Edit>();
        this.gop = new ArrayList<Packet>();
    }

    public void addTimecode(Packet packet) throws IOException {
        if(_timescale == NO_TIMESCALE_SET)
            _timescale = packet.getTimescale();
        if(_timescale != NO_TIMESCALE_SET && _timescale != packet.getTimescale())
            throw new RuntimeException("MP4 timecode track doesn't support timescale switching.");
        if (packet.isKeyFrame())
            processGop();
        gop.add(Packet.createPacketWithData(packet, (ByteBuffer) null));
    }

    private void processGop() throws IOException {
        if (gop.size() > 0) {
            for (Packet pkt : sortByDisplay(gop)) {
                addTimecodeInt(pkt);
            }
            gop.clear();
        }
    }

    private List<Packet> sortByDisplay(List<Packet> gop) {
        ArrayList<Packet> result = new ArrayList<Packet>(gop);
        Collections.sort(result, new Comparator<Packet>() {
            public int compare(Packet o1, Packet o2) {
                if (o1 == null && o2 == null)
                    return 0;
                else if (o1 == null)
                    return -1;
                else if (o2 == null)
                    return 1;
                else
                    return o1.getDisplayOrder() > o2.getDisplayOrder() ? 1 : (o1.getDisplayOrder() == o2
                            .getDisplayOrder() ? 0 : -1);
            }
        });
        return result;
    }

    protected Box finish(MovieHeaderBox mvhd) throws IOException {
        processGop();
        outTimecodeSample();

        if (sampleEntries.size() == 0)
            return null;

        if (edits != null) {
            edits = Util.editsOnEdits(new Rational(1, 1), lower, edits);
        } else
            edits = lower;

        return super.finish(mvhd);
    }

    private void addTimecodeInt(Packet packet) throws IOException {
        TapeTimecode tapeTimecode = packet.getTapeTimecode();
        boolean gap = isGap(prevTimecode, tapeTimecode);
        prevTimecode = tapeTimecode;

        if (gap) {
            outTimecodeSample();
            firstTimecode = tapeTimecode;
            fpsEstimate = tapeTimecode.isDropFrame() ? 30 : -1;
            samplePts += sampleDuration;
            sampleDuration = 0;
            tcFrames = 0;
        }
        sampleDuration += packet.getDuration();
        tcFrames++;
    }

    private boolean isGap(TapeTimecode prevTimecode, TapeTimecode tapeTimecode) {
        boolean gap = false;

        if (prevTimecode == null && tapeTimecode != null) {
            gap = true;
        } else if (prevTimecode != null) {
            if (tapeTimecode == null)
                gap = true;
            else {
                if (prevTimecode.isDropFrame() != tapeTimecode.isDropFrame()) {
                    gap = true;
                } else {
                    gap = isTimeGap(prevTimecode, tapeTimecode);
                }
            }
        }
        return gap;
    }

    private boolean isTimeGap(TapeTimecode prevTimecode, TapeTimecode tapeTimecode) {
        boolean gap = false;
        int sec = toSec(tapeTimecode);
        int secDiff = sec - toSec(prevTimecode);
        if (secDiff == 0) {
            int frameDiff = tapeTimecode.getFrame() - prevTimecode.getFrame();
            if (fpsEstimate != -1)
                frameDiff = (frameDiff + fpsEstimate) % fpsEstimate;
            gap = frameDiff != 1;
        } else if (secDiff == 1) {
            if (fpsEstimate == -1) {
                if (tapeTimecode.getFrame() == 0)
                    fpsEstimate = prevTimecode.getFrame() + 1;
                else
                    gap = true;
            } else {
                int firstFrame = tapeTimecode.isDropFrame() && (sec % 60) == 0 && (sec % 600) != 0 ? 2 : 0;

                if (tapeTimecode.getFrame() != firstFrame || prevTimecode.getFrame() != fpsEstimate - 1)
                    gap = true;
            }
        } else {
            gap = true;
        }
        return gap;
    }

    private void outTimecodeSample() throws IOException {
        if (sampleDuration > 0) {
            if (firstTimecode != null) {
                if (fpsEstimate == -1)
                    fpsEstimate = prevTimecode.getFrame() + 1;
                TimecodeSampleEntry tmcd = TimecodeSampleEntry.createTimecodeSampleEntry((firstTimecode.isDropFrame() ? 1 : 0), _timescale, (int) (sampleDuration / tcFrames), fpsEstimate);
                sampleEntries.add(tmcd);
                ByteBuffer sample = ByteBuffer.allocate(4);
                sample.putInt(toCounter(firstTimecode, fpsEstimate));
                sample.flip();
                addFrame(MP4Packet.createMP4Packet(sample, samplePts, _timescale, sampleDuration, 0, true, null, 0, samplePts, sampleEntries.size() - 1));

                lower.add(new Edit(sampleDuration, samplePts, 1.0f));
            } else {
                lower.add(new Edit(sampleDuration, -1, 1.0f));
            }
        }
    }

    private int toCounter(TapeTimecode tc, int fps) {
        int frames = toSec(tc) * fps + tc.getFrame();
        if (tc.isDropFrame()) {
            long D = frames / 18000;
            long M = frames % 18000;
            frames -= 18 * D + 2 * ((M - 2) / 1800);
        }

        return frames;
    }

    private int toSec(TapeTimecode tc) {
        return tc.getHour() * 3600 + tc.getMinute() * 60 + tc.getSecond();
    }
}

File: src/main/java/org/jcodec/containers/mp4/muxer/TimecodeMP4MuxerTrack.kt
package org.jcodec.containers.mp4.muxer

import org.jcodec.common.Codec
import org.jcodec.common.model.Packet
import org.jcodec.common.model.Packet.FrameType
import org.jcodec.common.model.Rational
import org.jcodec.common.model.TapeTimecode
import org.jcodec.containers.mp4.MP4Packet
import org.jcodec.containers.mp4.MP4TrackType
import org.jcodec.containers.mp4.boxes.Box
import org.jcodec.containers.mp4.boxes.Edit
import org.jcodec.containers.mp4.boxes.MovieHeaderBox
import org.jcodec.containers.mp4.boxes.TimecodeSampleEntry.Companion.createTimecodeSampleEntry
import org.jcodec.movtool.Util
import java.io.IOException
import java.nio.ByteBuffer
import java.util.*

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 *
 * Timecode MP4 muxer track
 *
 * @author The JCodec project
 */
class TimecodeMP4MuxerTrack(trackId: Int) : CodecMP4MuxerTrack(trackId, MP4TrackType.TIMECODE, Codec.TIMECODE) {
    private var prevTimecode: TapeTimecode? = null
    private var firstTimecode: TapeTimecode? = null
    private var fpsEstimate = 0
    private var sampleDuration: Long = 0
    private var samplePts: Long = 0
    private var tcFrames = 0
    private val lower: MutableList<Edit>
    private val gop: MutableList<Packet>

    @Throws(IOException::class)
    fun addTimecode(packet: Packet) {
        if (timescale == AbstractMP4MuxerTrack.NO_TIMESCALE_SET) timescale = packet.getTimescale()
        if (timescale != AbstractMP4MuxerTrack.NO_TIMESCALE_SET && timescale != packet.getTimescale()) throw RuntimeException("MP4 timecode track doesn't support timescale switching.")
        if (packet.isKeyFrame) processGop()
        gop.add(Packet.createPacketWithData(packet, null as ByteBuffer?))
    }

    @Throws(IOException::class)
    private fun processGop() {
        if (gop.size > 0) {
            for (pkt in sortByDisplay(gop)) {
                addTimecodeInt(pkt)
            }
            gop.clear()
        }
    }

    private fun sortByDisplay(gop: List<Packet>): List<Packet> {
        val result = ArrayList(gop)
        Collections.sort(result) { o1, o2 ->
            if (o1 == null && o2 == null) 0 else if (o1 == null) -1 else if (o2 == null) 1 else if (o1.getDisplayOrder() > o2.getDisplayOrder()) 1 else if (o1.getDisplayOrder() == o2
                            .getDisplayOrder()) 0 else -1
        }
        return result
    }

    @Throws(IOException::class)
    override fun finish(mvhd: MovieHeaderBox): Box? {
        processGop()
        outTimecodeSample()
        if (sampleEntries.orEmpty().size == 0) return null
        _edits = if (_edits != null) {
            Util.editsOnEdits(Rational(1, 1), lower, _edits)
        } else lower
        return super.finish(mvhd)
    }

    @Throws(IOException::class)
    private fun addTimecodeInt(packet: Packet) {
        val tapeTimecode = packet.getTapeTimecode()
        val gap = isGap(prevTimecode, tapeTimecode)
        prevTimecode = tapeTimecode
        if (gap) {
            outTimecodeSample()
            firstTimecode = tapeTimecode
            fpsEstimate = if (tapeTimecode.isDropFrame) 30 else -1
            samplePts += sampleDuration
            sampleDuration = 0
            tcFrames = 0
        }
        sampleDuration += packet.getDuration()
        tcFrames++
    }

    private fun isGap(prevTimecode: TapeTimecode?, tapeTimecode: TapeTimecode?): Boolean {
        var gap = false
        if (prevTimecode == null && tapeTimecode != null) {
            gap = true
        } else if (prevTimecode != null) {
            gap = if (tapeTimecode == null) true else {
                if (prevTimecode.isDropFrame != tapeTimecode.isDropFrame) {
                    true
                } else {
                    isTimeGap(prevTimecode, tapeTimecode)
                }
            }
        }
        return gap
    }

    private fun isTimeGap(prevTimecode: TapeTimecode, tapeTimecode: TapeTimecode): Boolean {
        var gap = false
        val sec = toSec(tapeTimecode)
        val secDiff = sec - toSec(prevTimecode)
        if (secDiff == 0) {
            var frameDiff = tapeTimecode.frame - prevTimecode.frame
            if (fpsEstimate != -1) frameDiff = (frameDiff + fpsEstimate) % fpsEstimate
            gap = frameDiff != 1
        } else if (secDiff == 1) {
            if (fpsEstimate == -1) {
                if (tapeTimecode.frame.toInt() == 0) fpsEstimate = prevTimecode.frame + 1 else gap = true
            } else {
                val firstFrame = if (tapeTimecode.isDropFrame && sec % 60 == 0 && sec % 600 != 0) 2 else 0
                if (tapeTimecode.frame.toInt() != firstFrame || prevTimecode.frame.toInt() != fpsEstimate - 1) gap = true
            }
        } else {
            gap = true
        }
        return gap
    }

    @Throws(IOException::class)
    private fun outTimecodeSample() {
        if (sampleDuration > 0) {
            if (firstTimecode != null) {
                if (fpsEstimate == -1) fpsEstimate = prevTimecode!!.frame + 1
                val tmcd = createTimecodeSampleEntry(if (firstTimecode!!.isDropFrame) 1 else 0, timescale, (sampleDuration / tcFrames).toInt(), fpsEstimate)
                sampleEntries?.add(tmcd)
                val sample = ByteBuffer.allocate(4)
                sample.putInt(toCounter(firstTimecode!!, fpsEstimate))
                sample.flip()
                addFrame(MP4Packet.createMP4Packet(sample, samplePts, timescale, sampleDuration, 0, FrameType.KEY, null, 0, samplePts, sampleEntries.orEmpty().size - 1))
                lower.add(Edit(sampleDuration, samplePts, 1.0f))
            } else {
                lower.add(Edit(sampleDuration, -1, 1.0f))
            }
        }
    }

    private fun toCounter(tc: TapeTimecode, fps: Int): Int {
        var frames = toSec(tc) * fps + tc.frame
        if (tc.isDropFrame) {
            val D = frames / 18000.toLong()
            val M = frames % 18000.toLong()
            frames -= (18 * D + 2 * ((M - 2) / 1800).toInt()).toInt()
        }
        return frames
    }

    companion object {
        private fun toSec(tc: TapeTimecode): Int {
            return tc.hour * 3600 + tc.minute * 60 + tc.second
        }
    }

    init {
        lower = ArrayList()
        gop = ArrayList()
    }
}